File: star_seller.ml
(* The Star Seller exercise in algorithmics,
   written in OCaML.                        [14 lines of code]
 *)

(* OCaML is a functional programming language, sure, but it's not pure.
   Coming from Haskell, type safety is cumbersome (you need to tie it into your
   function definition, instead of its own line) and what type there is is
   awkward (No type classes or traits means a different operator for ints and
   floats).

   I'm sure it has its advantages -- I mean, there's a reason why it is the
   preferred language for Mirage OS -- but it just leaves a bitter aftertaste
   as a Haskeller...
*)

let putStars n = Array.fold_left (^) "" (Array.make n "*") ;;

(* Infix operators are defined the same as in Haskell *)
let (<=>) x (a,b) = x >= a && x <= b;;

(* In order to be applicative (lazy), a definition needs at least one argument.
   Functions that do not take arguments in OCaML usually take a 'unit', or null
   typed value. *)
let rec askStars () =   (* OCaML functions aren't recursive by default *)
                let _ = Printf.printf "%s" "How many stars do you want? " in
                let result = int_of_string (read_line ()) in
                match result with
                | x when x <=> (0,200) -> x
                | _ -> askStars ();;


(* OCaML is imperative, meaning evaluation and execution follow the same flow,
   discarding the need for a 'main' function, which Haskell requires. *)

Printf.printf "%s" "What is your name? ";;
let name = read_line ();;

Printf.printf "Hello, %s!\n" name;;
let stars = askStars ();;

Printf.printf "%s\n" (putStars stars);;
Printf.printf "Goodbye, %s\n" name;;

(* This language is case-sensitive. *)

File: star_seller.swift
/**
 * The Star Seller exercise in algorithmics,
 * written in Swift.                        [34 lines of code]
 *
 * An interesting language for sure. It feels high-level yet natively handles
 * C types and function calls.
 */

// Interestingly, function parameters are named by default in Swift, and their
// label can differ from their internal variable name.
// Here, to be able to omit the label altogether, it is set to `_`
func putStars(_ stars: Int) {
    if (stars > 0) {
        for _ in 1...stars {
            print("*", terminator: "")
        }
    }
    print()
}

func star_seller() {
    // Swift, not unlike Rust, relies heavily on optional types. In fact,
    // a variable can only be nil if it is of an optional type.
    // Unwrapping is done using the `!` suffix.
    var name: String?
    var nstars: Int?
    var valid: Bool = false

    while (!valid) {
        print("What is your name?", terminator: " ")
        name = readLine()
        if (name != nil && name!.count > 0) {
            valid = true
        }
    }
    valid = false
    print("Hello, \(name!)!")
    while (!valid) {
        print("How many stars do you want?", terminator: " ")
        // Swift also supports using optionals in an if/else block, essentially
        // acting as a try/catch block of sorts.
        // The `let` keyword defines constants as opposed to `var` which
        // defines variables.
        if let stars = readLine() {
            nstars = Int(stars)
        }
        if (nstars != nil && nstars! <= 200 && nstars! >= 0) {
            putStars(nstars!)
            valid = true
        }
    }
    print("Goodbye, \(name!)")
}

// There is no explicit entry point in Swift, not unlike Python.
star_seller()

// This language is case-sensitive.
